h2o-core/src/main/java/water/udf/DkvClassLoader.java

```diff
@@ -28,20 +38,138 @@ public DkvClassLoader(String jarKeyName, ClassLoader parent) {
-    Value v = water.DKV.get(jarKey);
-    // Get local persistent layer and use it to save
-    // content of K/V
-    PersistFS persistFS = (PersistFS) H2O.getPM().getIce();
-    File f = persistFS.getFile(v);
-    // Optimistic delete
-    f.deleteOnExit();
+    super(parent);
+    this.jarKey = jarKey;
+    this.jarCache = buildJarCache(jarKey);
+  }
+  
+  @Override
+  protected Class<?> findClass(String name) throws ClassNotFoundException {
     try {
-      Log.debug("DkvClassLoader: saving " + v + " into " + f.getAbsolutePath());
-      H2O.getPM().store(Value.ICE, v);
-      addURL(f.toURI().toURL());
-    } catch (Exception e) {
-      throw new IllegalArgumentException(e);
+      return super.findClass(name);
+    } catch (ClassNotFoundException e) {
+      // Parent does not contain the requested class, look into cache we built.
+      String path = name.replace('.', '/').concat(".class");
+      byte[] klazzBytes = jarCache.get(path);
+      if (klazzBytes != null && klazzBytes.length > 0) {
+        return defineClass(name, klazzBytes, 0, klazzBytes.length);
+      }
+
+      throw new ClassNotFoundException(name);
+    }
+  }
+  @Override
+  protected URL findResource(String name) {
+    return url(name);
+  }
+
+  @Override
+  protected Enumeration<URL> findResources(String name) {
+    URL url = url(name);
+    return url == null
+           ? Collections.<URL>emptyEnumeration()
+           : Collections.enumeration(Collections.singletonList(url));
+  }
+
+  protected URL url(String name) {
+    URL url = null;
+    byte[] content = jarCache.get(name);
+    if (content != null) {
+      try {
+        // Create a nice URL representing the resource following concept of JarUrl.
+        // Note: this is just for sake of clarity, but at this
+        // point we use cached content to serve content of URL.
+        url = new URL(PROTO, "", -1,
+                      this.jarKey + (name.startsWith("/") ? "!" : "!/") + name,
+                      new DkvUrlStreamHandler());
+      } catch (MalformedURLException e) {
+        // Fail quickly since this is not expected to fail
+        throw new RuntimeException(e);
+      }
+    }
+    return url;
+  }
+
+  static Map<String, byte[]> buildJarCache(Key jarKey) {
+    Map<String, byte[]> jarCache = new HashMap<>();
+    try(JarInputStream jis = new JarInputStream(new ByteArrayInputStream(DKV.get(jarKey).memOrLoad()))) {
+      JarEntry entry = null;
+      while ((entry = jis.getNextJarEntry()) != null) {
+        if (entry.isDirectory()) continue;
+        byte[] content = readJarEntry(jis, entry);
+        jarCache.put(entry.getName(), content);
+      }
+    } catch (IOException e) {
+      // Fail quickly
+      throw new RuntimeException(e);
+    }
+    return jarCache;
+  }
+
+  static byte[] readJarEntry(JarInputStream jis, JarEntry entry) throws IOException {
+    int len = (int) entry.getSize();
+    byte[] content = new byte[len > 0 ? len : 4096];
```

Comment:
Can you use IOUtils.toByteArray? Or do you not like that the writes are synchronized in ByteArrayOutputStream?

Reply:
You mean IOUtils.toByteArray(entry)?

Reply:
It should work with `JarInputStream`, that is 

    byte[] content = len >= 0 ? IOUtils.toByteArray(jis, len) : IOUtils.toByteArray(jis)

Reply:
nice!